sources/pods.go

```diff
@@ -0,0 +1,135 @@
+type podsApi interface {
+	// Returns a list of pods that exist on the nodes in 'nodeList'
+	List(nodeList *nodes.NodeList) ([]Pod, error)
+
+	// Returns debug information.
+	DebugInfo() string
+}
+
+type realPodsApi struct {
+	client *client.Client
+	// a means to list all scheduled pods
+	podLister *cache.StoreToPodLister
+}
+
+func (self *realPodsApi) parsePod(pod *api.Pod) *Pod {
+	localPod := Pod{
+		Name:       pod.Name,
+		Namespace:  pod.Namespace,
+		ID:         string(pod.UID),
+		Hostname:   pod.Status.Host,
+		HostIP:     pod.Status.HostIP,
+		Status:     string(pod.Status.Phase),
+		PodIP:      pod.Status.PodIP,
+		Labels:     make(map[string]string, 0),
+		Containers: make([]*Container, 0),
+	}
+	for key, value := range pod.Labels {
+		localPod.Labels[key] = value
+	}
+	for _, container := range pod.Spec.Containers {
+		localContainer := newContainer()
+		localContainer.Name = container.Name
+		localPod.Containers = append(localPod.Containers, localContainer)
+	}
+	glog.V(3).Infof("parsed kube pod: %+v", localPod)
+
+	return &localPod
+}
+
+func (self *realPodsApi) parseAllPods(pods []api.Pod) []Pod {
+	out := make([]Pod, 0)
+	for _, pod := range pods {
+		glog.V(3).Infof("Found kube Pod: %+v", pod)
+		pod := self.parsePod(&pod)
+		out = append(out, *pod)
+	}
+
+	return out
+}
+
+func (self *realPodsApi) getNodeSelector(nodeList *nodes.NodeList) (labels.Selector, error) {
+	nodeLabels := []string{}
+	for node := range nodeList.Items {
+		nodeLabels = append(nodeLabels, fmt.Sprintf("Status.Host==%s", node.Name))
+	}
+	glog.V(2).Infof("using labels %v to find pods", nodeLabels)
+	return labels.ParseSelector(strings.Join(nodeLabels, ","))
+}
+
+// Returns a map of minion hostnames to the Pods running in them.
+func (self *realPodsApi) List(nodeList *nodes.NodeList) ([]Pod, error) {
+	pods, err := self.podLister.List(labels.Everything())
+	if err != nil {
+		return []Pod{}, err
+	}
+	selectedPods := []api.Pod{}
+	// TODO(vishh): Avoid this loop by setting a node selector on the watcher.
+	for _, pod := range pods {
+		if _, ok := nodeList.Items[nodes.Node{pod.Status.Host, pod.Status.HostIP}]; ok {
+			selectedPods = append(selectedPods, pod)
+		}
+	}
+	glog.V(2).Infof("got pods from api server %+v", selectedPods)
+
+	return self.parseAllPods(pods), nil
+}
+
+func (self *realPodsApi) DebugInfo() string {
+	return ""
+}
+
+func (self *realPodsApi) createAssignedPodLW() *cache.ListWatch {
+	selector, err := labels.ParseSelector("Status.Host!=")
+	if err != nil {
+		panic(err)
+	}
+
+	return &cache.ListWatch{
+		Client:        self.client,
+		FieldSelector: selector,
+		Resource:      "pods",
+	}
+}
+
+func newPodsApi(client *client.Client) podsApi {
+	podsApi := &realPodsApi{
+		client:    client,
+		podLister: &cache.StoreToPodLister{cache.NewStore(cache.MetaNamespaceKeyFunc)},
+	}
+	// Watch and cache all running pods.
+	cache.NewReflector(podsApi.createAssignedPodLW(), &api.Pod{}, podsApi.podLister.Store).Run()
```

Comment:
Yeah. We poll on a in-memory cache that gets updated asynchronously. In the future we can move to a model where we identify creation and deletion of pods and have a handler per pod. I will have to dig into the watch implementation for that.